### ISSUE TITLE
Project deletion on PR closed event

### ISSUE THREAD

**Issue by milo-minderbinder** [checkmarx-ltd/cx-flow#556]
### Describe the problem
For GitHub integrations (possibly others as well), it is only possible to configure automatic project deletion in CxFlow on branch/tag deletion events. For teams using a fork based git flow (only create PRs from forked repos), however, this means that PR projects will never be deleted, since branch deletion event is never sent for forks. Even if these teams try to work around that by installing the CxFlow webhook in their forked repos, branch deletion event will not be processed for the intended Checkmarx project unless the team and project name is overridden using config or groovy scripts, which even then is very difficult since the deletion event data does not include any references to the original PR.
Also (though less important) — some teams do not want to keep Checkmarx projects that came from PRs/branches that required additional work and the PR was closed.

### Proposed solution
Optionally support project deletion on pull request closed events instead of, or in addition to, on branch deletion. Incidentally, doing so makes almost all of the issues with flows based on forks go away.

### Additional details
Here’s a patch to add deletion on PR closure as a reference. It’s been tested and is used in our prod deployment successfully, but if this feature request is accepted this should probably be a configurable option (and patch could probably use some cleaning up/refactoring).

```diff
From 57e02da726e4cd23866b22185e6639bd02563d39 Mon Sep 17 00:00:00 2001
From: Milo Minderbinder <minderbinder.enterprises@gmail.com>
Date: Thu, 3 Dec 2020 00:23:39 -0500
Subject: [PATCH] delete Checkmarx project on PR close

---
 .../flow/controller/GitHubController.java      | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/src/main/java/com/checkmarx/flow/controller/GitHubController.java b/src/main/java/com/checkmarx/flow/controller/GitHubController.java
index df9e8328..6bf0a903 100644
--- a/src/main/java/com/checkmarx/flow/controller/GitHubController.java
+++ b/src/main/java/com/checkmarx/flow/controller/GitHubController.java
@@ -125,7 +125,8 @@ public class GitHubController extends WebhookController {
             String action = event.getAction();
             if(!action.equalsIgnoreCase("opened") &&
                     !action.equalsIgnoreCase("reopened") &&
-                    !action.equalsIgnoreCase("synchronize")){
+                    !action.equalsIgnoreCase("synchronize") &&
+                    !action.equalsIgnoreCase("closed")){
                 log.info("Pull requested not processed.  Status was not opened ({})", action);
                 return ResponseEntity.status(HttpStatus.OK).body(EventResponse.builder()
                         .message("No processing occurred for updates to Pull Request")
@@ -224,7 +225,20 @@ public class GitHubController extends WebhookController {
             request.setId(uid);
             //only initiate scan/automation if target branch is applicable
             if(helperService.isBranch2Scan(request, branches)){
-                flowService.initiateAutomation(request);
+                if (action.equalsIgnoreCase("closed")) {
+
+                    flowService.deleteProject(request);
+
+                    final String MESSAGE = "PR closed event was handled successfully.";
+                    log.info(MESSAGE);
+
+                    return ResponseEntity.status(HttpStatus.OK).body(EventResponse.builder()
+                            .message(MESSAGE)
+                            .success(true)
+                            .build());
+                } else {
+                    flowService.initiateAutomation(request);
+                }
             }
         } catch (IllegalArgumentException e) {
             return getBadRequestMessage(e, controllerRequest, product);
-- 
2.29.2
```

**Comment by kmcdon83** [comment]
@jbrotsos , perhaps we perform a check that a project exists for the "main" or "master" branch before performing the delete.

@milo-minderbinder , the main concern we have with this workflow is licensing.  If you only scanned on PR, we would create and delete the project at all times.  We would need to ensure at lest 1 <REPO>-* project exists to avoid license abuse.

**Comment by milo-minderbinder** [comment]
@jbrotsos any update on this? I don’t like having to maintain a forked version of CxFlow and the license concerns mentioned above are not relevant anymore with ([branching](https://github.com/checkmarx-ltd/cx-flow/wiki/Configuration#branched-project) enabled, since CxFlow already creates a base project for the default branch if one does not already.

I’d also suggest that when using GitHub app based integration, CxFlow should handle the installation deleted and installation-repositories remove repos events by deleting any previously created projects for those repos where the app was removed.

**Comment by itsKedar** [comment]
Fixed in 1.7.02
